engine/src/main/java/org/hibernate/search/engine/search/projection/spi/ArrayProjectionCollector.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.search.engine.search.projection.spi;
+
+import java.lang.reflect.Array;
+import java.util.List;
+
+import org.hibernate.search.engine.search.projection.ProjectionCollector;
+
+/**
+ * A {@link ProjectionCollector} that can accumulate any number of values into a {@code V[]}.
+ *
+ * @param <E> The type of extracted values to accumulate before being transformed.
+ * @param <V> The type of values to accumulate obtained by transforming extracted values ({@code E}).
+ */
+final class ArrayProjectionCollector<E, V> extends ListBasedProjectionCollector<E, V, V[]> {
+
+	static <U, R> Provider<U, R> provider(Class<? super U> elementType) {
+		return new ArrayProvider<>( elementType );
+	}
+
+	private ArrayProjectionCollector(Class<? super V> elementType) {
+		this.elementType = elementType;
+	}
+
+	private final Class<? super V> elementType;
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public V[] doFinish(List<V> accumulated) {
+		V[] array = (V[]) Array.newInstance( elementType, accumulated.size() );
```

Comment:
Soooo following the discussion at https://hibernate.zulipchat.com/#narrow/channel/132094-hibernate-orm-dev/topic/StandardStack.20and.20reflection
I guess this won't fly well with Quarkus too 😖 😃....

I can change this to use `IntFunction<T[]>` and simply use `accumulated.toList( ... )` but then..... it won't work fine here:

https://github.com/hibernate/hibernate-search/blob/12cd6cb0fefbcd068285a73daef5d2635fb067ff/mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/search/definition/binding/impl/ProjectionBindingContextImpl.java#L417-L419

so we could do something like for sorted sets and throw an exception. Or give two ways to create array collector with class and with function. Use class in the `ProjectionBindingContextImpl` and it'll fail for users so they'd need to register their arrays themselves .... @yrodiere WDYT ? 🙈

Reply:
I think it's fine-ish, as this only happens when a user explicitly requests an array, so it's reasonable to expect them to register that array type for reflection.

For `ProjectionBindingContextImpl`, I suppose we can add something to Quarkus that would detect projection constructors and register for reflection any array type used in their parameters. But only if we consider this important enough -- I personally don't. I wouldn't even have provided the feature TBH, because using arrays in domain models is a big no-no in my books. But now that it's there, I guess we might as well keep it.